### ISSUE TITLE
Feature request: Mermaid (Markdown)

### ISSUE THREAD

**Issue by mtompkins** [thorsten/phpMyFAQ#1091]
**Mermaid** - with the fantastic move to MarkDown, I'd love to see this utility incorporated. I haven't seen mention of it anywhere so I'm adding it here. https://github.com/knsv/mermaid It is SO convenient especially in a FAQ system

**Comment by thorsten** [comment]
It's just a JS library as far as I see, right?

**Comment by mtompkins** [comment]
Yes, and NPM friendly which should integrate nicely (and hopefully seemlessly) for you.

**Comment by thorsten** [comment]
Should be easy to add that even to 2.9

**Comment by mtompkins** [comment]
I was thinking it should be a very quick and solid addition for phpmyfaq (and of course the user base) based on the addition of Markdown. Hopefully that proves to be the case.

**Comment by thorsten** [comment]
Will be added in the upcoming RC2 release.

**Comment by mtompkins** [comment]
Took RC2 for a spin just now - really looks polished and amazing! Congratulations!

Just a note / followup on mermaid - it doesn't seem to be incorporated in RC2. Perhaps you meant RC3?

Here is some test Markdown to reproduce / test mermaid (which will not render properly on Github)
![screenshot from 2016-03-20 07-55-38](https://cloud.githubusercontent.com/assets/4105910/13904284/7a2353f4-ee71-11e5-8ccf-ebe1e8fe6318.png)
:
<---------- BEGIN ---------->

# Test

```
test
```

```
#!/bin/bash
TEST=9
```
~~~cpp
TEST=9
~~~

~~~mermaid
graph LR;
A-->B;
A-->C;
B-->D;
C-->D;
~~~

~~~mermaid
graph TD;
A-->B;
A-->C;
B-->D;
C-->D;
~~~

~~~mermaid
graph LR;
A[Hard edge]-->|Link text|B(Round edge);
B-->C{Decision}
C-->|One|D[Result one];
C-->|Two|E[Result two];
~~~
~~~mermaid
gantt
        dateFormat  YYYY-MM-DD
        title Adding GANTT diagram functionality to mermaid
        section A section
        Completed task            :done,    des1, 2014-01-06,2014-01-08
        Active task               :active,  des2, 2014-01-09, 3d
        Future task               :         des3, after des2, 5d
        Future task2               :         des4, after des3, 5d
        section Critical tasks
        Completed task in the critical line :crit, done, 2014-01-06,24h
        Implement parser and jison          :crit, done, after des1, 2d
        Create tests for parser             :crit, active, 3d
        Future task in critical line        :crit, 5d
        Create tests for renderer           :2d
        Add to mermaid                      :1d
~~~
# Heading 1
to
###### Heading 6

<----------- END ------------->

**Comment by thorsten** [comment]
Hm, strange. Does it need some configuration? Just added the library

**Comment by mtompkins** [comment]
Here is the relevant doc: http://knsv.github.io/mermaid/#installation

Presume you are doing:
```
<script src="mermaid.min.js"></script>
<script>mermaid.initialize({startOnLoad:true});</script>
```

**Comment by mtompkins** [comment]
Further reading seems to indicate best way forward is a call to the API:

http://knsv.github.io/mermaid/#api-usage

**Comment by thorsten** [comment]
Okay, I forgot that :-(
Sorry, preparing a 3rd RC asap

**Comment by mtompkins** [comment]
Not a problem ... that's why it's an RC!! Will test when available.

**Comment by thorsten** [comment]
Added these lines, but it still don't work...

**Comment by mtompkins** [comment]
I see your approach was to enable mermaid only if markdown is active. Shall I suggest we start with enable mermaid at all times (it works with html directly) as a place to start?

The following HTML should generate a simple diagram:

```HTML
<div class="mermaid">
        graph LR
            A-->B
            B-->C
            C-->A
            D-->C
</div>
```

**Comment by thorsten** [comment]
Right, shouldn't have any side effects. One thing that seems to break the code is that the arrows are encoded...

**Comment by mtompkins** [comment]
I've seen something similar on one of the downstream projects (arrow / symbol issue). I'll see if I can find it, test & revert.

**Comment by mtompkins** [comment]
Which markdown to HTML renderer are you using?

**Comment by mtompkins** [comment]
Here are the encoding (arrow) references I was remembering:
+ http://mostlyblather.com/blog/2015/05/23/mermaid-jekyll-octopress/
+ http://nauvalatmaja.com/2015/01/13/rendering-mermaid-in-docpad/

**Comment by mtompkins** [comment]
OK - here's where I am and I think we should pause...there is an upstream issue.

What I've done is relocate the instantiation of mermaid from artikel.php to index.tpl on the bottom as is often used for `<script>` instantiation. Also as mermaid works with or without markdown (i.e. standard HTML works find to produce a graphic) I believe it should be always enabled.

Doing that minor edit I'm left with this error:
```
No longer working - Invalid value for <svg> attribute viewBox="0 0 -Infinity -Infinity"
```

Here are the relevant issues on upstream channels:
+ https://github.com/knsv/mermaid/issues/291
+ https://github.com/jasonbellamy/jekyll-mermaid/issues/5

Once we see how these are resolved I'll pick it back up. Once we render the graphics correctly I'll see about the direct implementation to the markdown editor.

If you are using `marked` as the MD --> HTML engine we can easily handle the arrows based on the references here: https://github.com/thorsten/phpMyFAQ/issues/1091#issuecomment-198963193

**Comment by thorsten** [comment]
Okay, I'll postpone this feature to 2.9.1.

**Comment by thorsten** [comment]
btw, we use Parsedown: http://parsedown.org/

**Comment by mtompkins** [comment]
Possibly related to https://github.com/thorsten/phpMyFAQ/issues/1098 in phpMyFAQ use case as newline characters are not being correctly captured and streamed in the editor to be parsed.

**Comment by mtompkins** [comment]
OK - some progress made. All indications are that because you are concatenating the mermaid.js, you are calling the initialization of the function before it is defined in your previous attempt. We need the FAQ article to have been rendered in totality before calling this function. Please consider where to put that initialization which should be of the form:`mermaid.initialize();` On a simplied webpage this would go in a `document.ready()` event.

The following code enetered in the **Markdown** editor should be used to test:
```html
<div class="mermaid">
graph TD;
A-->B;
A-->C;
B-->D;
C-->D;
</div>
```

This will generate the the initial diagram as shown:
![screenshot from 2016-03-21 09-59-38](https://cloud.githubusercontent.com/assets/4105910/13920899/ab174170-ef4b-11e5-8efe-3b2c38231c57.png)

The call to mermaid should parse the connectors `-->`

**Comment by thorsten** [comment]
I'm still getting the SVG error in the console, but I see an empty SVG in the developer tools. Would it be possible to add a pull request with your working result?

**Comment by mtompkins** [comment]
It isn't working at the moment. It isn't making any call to mermaid until we add that initialization which I'm waiting for your guidance on where to add.

**Comment by mtompkins** [comment]
Perhaps this will help in understanding. When I look at the source of a FAQ, I see the following:
```
<script>

            $(function() {
                $('div.pmf-stars > div.pmf-star-rating > span').on('click', function(e) {
                    var numStars = $(e.target).data('stars');
                    saveVoting('faq', 1, numStars, 'en-us');
                });
            });
            $('form#formValues').on('submit', function (e) {
                e.preventDefault();
                saveFormValues('savecomment', 'comment');
                return false;
            });
            $(document).ready(function() {
                hljs.configure({
                   useBR: true
                });
                $('pre code').each(function(i, block) {
                    hljs.highlightBlock(block);
                });
            });
        </script>
```

I'm not quite sure how this is generated. This is likely the appropriate place to insert that initialization.

Can you advise where this is within the code or how it is injected?
Ideally I'd like to try:
```
<script>

            $(function() {
                $('div.pmf-stars > div.pmf-star-rating > span').on('click', function(e) {
                    var numStars = $(e.target).data('stars');
                    saveVoting('faq', 1, numStars, 'en-us');
                });
            });
            $('form#formValues').on('submit', function (e) {
                e.preventDefault();
                saveFormValues('savecomment', 'comment');
                return false;
            });
            $(document).ready(function() {
                mermaid.initialize();
                hljs.configure({
                   useBR: true
                });
                $('pre code').each(function(i, block) {
                    hljs.highlightBlock(block);
                });
            });
        </script>
```

**Comment by mtompkins** [comment]
OK - I found it in the tpl file. It seems Github (and specifically Github search) was having some difficulties

**Comment by thorsten** [comment]
The code you posted is located in phpmyfaq/assets/template/default/artikel.tpl

**Comment by mtompkins** [comment]
Almost there. Just have to decide why the arrow heads aren't being rendered.
Here's the help I need - right now I'm using added to index.tpl:
```
<link rel="stylesheet" href="https://cdn.rawgit.com/knsv/mermaid/0.5.6/dist/mermaid.css">
```
Instead we should use the local mermaid.css file that comes with the package. Can you tell me how to make that change?

Here's a photo of where we are (notice clean / no errors in console):
![screenshot from 2016-03-21 16-00-47](https://cloud.githubusercontent.com/assets/4105910/13932480/8d69818c-ef7e-11e5-8b24-06832ba5ee3f.png)

**Comment by mtompkins** [comment]
So in a short recap of how to reproduce where I am - a git pull is overkill.
Steps are:
- Add mermaid (actually renable as you just commented out the module)
- Add the mermaid.css file to the CSS collection (I recommend the `forest.mermaid.css` file

That's actually all it takes. The arrow heads will take some researching and I get the sense it isn't a CSS thing but an upstream update. The module is initializing itself (after a bit of digging) through the js file. If we find we need to overload it, the proper place to do that is the `artikel.tpl` file under `document.ready()`

To use it just wrap any Mermaid syntax in a `<div class="mermaid"></div>` Just remember it must be in the **markdown** editor.

Once some dust settles I'll mock up a little bit of instruction for users.
In the meantime, syntax is here: http://knsv.github.io/mermaid/index.html#flowcharts-basic-syntax

I think you can safely add this to RC2 or you can wait for the next candidate.
I'll inquire with the Mermaid devloper and see if he has any ideas on the arrowhead.

**Comment by thorsten** [comment]
I pushed my current state, please check it as it doesn't work for me.

**Comment by mtompkins** [comment]
Reviewing now.
Please paste for me the code you are putting into the markdown editor just to confirm.

**Comment by thorsten** [comment]
`graph TD;
A-->B;
A-->C;
B-->D;
C-->D;`

**Comment by mtompkins** [comment]
Ok - on your code you need the div wrapping: 
```
<div class="mermaid">
graph TD;
A-->B;
A-->C;
B-->D;
C-->D;
</div>
```

Please delete: https://github.com/thorsten/phpMyFAQ/blob/2.9/phpmyfaq/assets/template/default/artikel.tpl#L173
(no initialization here necessary)

Otherwise I think we are the same. Still reviewing.

**Comment by mtompkins** [comment]
Recent push works fine for me.
My steps:
- git clone git://github.com/thorsten/phpMyFAQ.git
- cd phpMyFAQ
- git checkout 2.9
- curl -s https://getcomposer.org/installer | php
- php composer.phar update
- php composer.phar install
- npm install
- bower install
- grunt

- Copy phpmyfaq to www, mkdir attachments data images, chmod -R www-data
- Set up wizard
- Enable markdown editor
- Create FAQ

Copy and paste entire snippet in to markdown editor
```
<div class="mermaid">
graph TD;
A-->B;
A-->C;
B==>D;
C-->D;
</div>

<div class="mermaid">
graph LR;
    A-->B(B)
    A-->C>C]
    B-->D((D))
    C-->D
classDef eco fill:green;
class A,B,C,D eco
</div>

<div class="mermaid">
sequenceDiagram
    Alice->John: Hello John, how are you?
    John-->Alice: Great!
</div>

<div class="mermaid">
graph TD; A-->B; A-->C; B-->D; C-->D;
</div>
```
**Note:** It won't show in the preview. Go to the front end article.
![screenshot from 2016-03-21 17-45-39](https://cloud.githubusercontent.com/assets/4105910/13935256/c33e8cb8-ef8c-11e5-9ec4-a14dcd0bf99f.png)

**Comment by thorsten** [comment]
Ack, now it works! Awesome! It needs some documentation.

**Comment by mtompkins** [comment]
I filed a bug with the author and will advise when we connect. I think we should wait until I hear something before we widely make the feature known. It's there - it works - but those arrow heads really are important. The link to the website provides all the syntax necessary if someone is eager.

We can also think about replacing the `<div>` wrapper with something a little more user-friendly. As you use Parsedown I'm not familiar with how to override it (documentation is terrible I looked).

This is how it is done with **Marked** renderer. Perhaps it will give you an idea
```
var renderer = new marked.Renderer();
    renderer.code = function (code, language) {
        if(code.match(/^sequenceDiagram/)||code.match(/^graph/)){
            return '<div class="mermaid">'+code+'</div>';
        }
        else{
            return '<pre><code>'+code+'</code></pre>';
        }
    };
```

**Comment by mtompkins** [comment]
Please revert https://github.com/thorsten/phpMyFAQ/blob/2.9/phpmyfaq/artikel.php#L78
There is no need to add this wrapper.

**Comment by thorsten** [comment]
Okay, so these things are missing:

- [ ] end user documentation
- [ ] Mermaid in admin preview
- [ ] editor helper plus replacing the <div> wrapper with some more user friendly

**Comment by mtompkins** [comment]
Please see: https://github.com/knsv/mermaid/issues/330#issuecomment-199958968

In order to accurately answer questions the mermaid developer may ask, I think direct interaction between you and him is probably the best road forward. I'll monitor the exchange and offer any help I can.

**Comment by knsv** [comment]
Hello :)

Is there a public url to the page with the missing arrows? If so I can take a look and see if I see something that could cause the problem.

**Comment by thorsten** [comment]
Hi, I could install a current snapshot somewhere.

**Comment by knsv** [comment]
Sounds like a  plan!

**Comment by thorsten** [comment]
Installation will be up and running the next days, sorry for the delay.

**Comment by thorsten** [comment]
Sorry for the delay, here's the installation: 
http://jen.demo.phpmyfaq.de/

Username
    admin 
Password
    demoadmin

**Comment by thorsten** [comment]
Any news on this?

**Comment by thorsten** [comment]
Ne feedback anymore. :(

**Comment by knsv** [comment]
Sorry, on varation. Will look at this when I get back. 

Get Outlook for iOS




On Sun, Aug 7, 2016 at 7:57 PM +0200, "Thorsten Rinne" <notifications@github.com> wrote:












Ne feedback anymore. :(

**Comment by knsv** [comment]
I have looked at this now. Before we get to mermaid I must say that phpMyFAQ is getting to be an impressive publishing system, simple, elegant and versatile. Well done!

Now to the issue at hand. I think the problem is caused by the use of a base tag in phpMyFAQ. This messes up the references to the arrow heads in the generated svg markup. Fortunately there is a configuration option for this in mermaid. If you initialize mermaid with the option arrowMarkerAbsolute set to true it should fix the arrows. Then the references in the arrows heads will be absolute based in the url in the page. This is not always what you want if you for instance want to render the svg outside of the browser but in this case it should be ok.

```
    var config = {
        startOnLoad:true,
        arrowMarkerAbsolute:true
    };
    mermaid.initialize(config);
``` 

Try it out and let me know how it goes!

**Comment by thorsten** [comment]
Thanks a lot, fixed for 2.9.5!
